lib/cfme/cloud_services/payload_builder.rb

```diff
@@ -0,0 +1,107 @@
+require 'minitar'
+
+module Cfme
+  module CloudServices
+    module PayloadBuilder
+      BOUNDARY = "AaB03x".freeze
+
+      module RequestParams
+        UPLOADER_HOST = "localhost".freeze
+        UPLOADER_PORT = "8080".freeze
+        UPLOADER_URI = "/api/ingress/v1/upload".freeze
+
+        CONTENT_TYPE = "application/vnd.redhat.topological-inventory.something+tgz".freeze
+
+        def org_id
+          54_321 # TODO: tie with subscription
+        end
+
+        def tenant
+          12_345 # TODO: tie with subscription
+        end
+
+        def x_rh_insights_request_id
+          # TODO: generate?
+          # request_id: Unique ID provided to the payload created by 3Scale.
+          # This ID will be used for the life of the object.
+          "52df9f748eabcfea"
+        end
+
+        def identity_content
+          {"identity" => {"account_number" => tenant}, "internal" => {"org_id" => org_id}}
+        end
+
+        def identity_headers
+          {"x-rh-identity" => Base64.strict_encode64(JSON.dump(identity_content)) }
+        end
+
+        def header
+          {"Content-Type"          => "multipart/form-data;boundary=\"#{BOUNDARY}\"",
+           "x-rh-insights-request" => x_rh_insights_request_id}.merge(identity_headers)
+        end
+
+        def inventory_file_name
+          "cfme_inventory_#{tenant}_#{org_id}"
+        end
+      end
+
+      def create_tempfile(data, source, binary = false)
+        file = Tempfile.new(source)
+        file.binmode if binary
+        file.write(data)
+        file.rewind
+        file
+      end
+
+      def unlink_file(file)
+        if file
+          path_file = file.path
+          $log.warn "Unable to remove temporary file: #{path_file}" unless file.unlink
+          $log.warn "Temporary file #{path_file} has not been removed" if File.exist?(path_file)
+        end
+      end
+
+      def packed_file_from(gzipped_tar)
+        begin
+          binary_file = create_tempfile(gzipped_tar, inventory_file_name, binary = true)
+        ensure
+          begin
+            yield(binary_file.path)
+          ensure
+            binary_file&.close
+            unlink_file(binary_file)
```

Comment:
Pretty sure you can just do `binary_file&.unlink`

Reply:
Also, if you yielded the binary_file object you could handle the common ensure blocks which are duplicated here and [here](https://github.com/RedHatCloudForms/cfme-cloud_services/pull/7/files#diff-52156b755601198e173c36121053f0f1R85) in that `create_tempfile` method

Reply:
yes but as I used `Gem::Package::TarWriter` this is pretty changed. (there is no need to use `create_tempfile `)

Reply:
:heart: